README.md
# IntuneEnhancedInventory
Repository for the Intune Custom Inventory solution by MSEndpointmgr.com

> IMPORTANT! 
> Version 1.2 requires use of version 3.5.0 of the Invoke-CustomInventoryAzureFunction.ps1 to be used in Proactive Remediations
> This version of the Azure Function will work for any custom log you want to send securely to Log Analytics

### Version History 
Full changelog can be found here: [Changelog](https://github.com/MSEndpointMgr/IntuneEnhancedInventory/blob/main/CHANGELOG.MD)
#### Latest Version for the Azure Function 
* 1.2 - Released 15.10.2022 

#### Latest Version history for the Proactive Remediation Script
* 3.5 - Released 15.10.2022

# Update ONLY
* To perform an update use this deploy button and enter information from your current deployment-

[![Deploy to Azure](https://aka.ms/deploytoazurebutton)](https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FMSEndpointMgr%2FIntuneEnhancedInventory%2Fmain%2FDeploy%2FUpdate%2FUpdateSecuredEnhancedInventory.json)

# Installation 
## Option 1 (legacy and not maintained) 
Use the simple proactive remediation that sends data direct to Log Analytics Workspace with secrets in code. 
Read the blogpost: 
[https://msendpointmgr.com/2021/04/12/enhance-intune-inventory-data-with-proactive-remediations-and-log-analytics/](https://msendpointmgr.com/2021/04/12/enhance-intune-inventory-data-with-proactive-remediations-and-log-analytics/)

## Option 2 (strongly recommended)
Use the new and updated proactive remediation that sends data through a Azure Function App to keep secret out of code and secure that only approved and known clients can send data to your log workspace. 

1. Deploy Azure Function using our template.  
[![Deploy to Azure](https://aka.ms/deploytoazurebutton)](https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FMSEndpointMgr%2FIntuneEnhancedInventory%2Fmain%2FDeploy%2FSecuredEnhancedInventory.json) 
3. Set API Permissions for MSI to graph with Add-MSIGraphPermissions.ps1 
4. Deploy the Invoke-CustomInventoryAzureFunction.ps1 Proactive remediation after you added your Azure Function URL to the script. 
Read the blogpost: 
[https://msendpointmgr.com/2022/01/17/securing-intune-enhanced-inventory-with-azure-function/ ](https://msendpointmgr.com/2022/01/17/securing-intune-enhanced-inventory-with-azure-function/)

### Example code for adding a custom log
```powershell 
$LogPayLoad = New-Object -TypeName PSObject 
$LogPayLoad | Add-Member -NotePropertyMembers @{$LogName1 = $Logdata1}
$LogPayLoad | Add-Member -NotePropertyMembers @{$LogName2 = $Logdata2}
# Construct main payload to send to LogCollectorAPI
$MainPayLoad = [PSCustomObject]@{
	AzureADTenantID = $AzureADTenantID
	AzureADDeviceID = $AzureADDeviceID
	LogPayloads = $LogPayLoad
}
...


CHANGELOG.MD
# Changelog for Intune Enhanced Inventory Solution 

## Changelog for Log Collector API Azure Function

### 1.2 
* Requires Proactive Remeditaion Script version 3.5
* Changed response logic in the function. Instead of returning a concated string, the function now returns a payload containing logname and status for each log as a separate entry. This makes it much easier to validate each log ingestion separated if you have many logs in the same PR Script. Fixes issues with the PR reporting wrongly. 
* Function now checks if the log you are sending is not empty. If the log is empty the function will not try to send data to Log Analytics and will return status 200 with a message about log empty. This fixes that the PR returns and error if you are collection for instance specific eventlog entries and the last 24 hours contains zero entries. 

### 1.1 
* Requires Proactive Remediation Script version 2.0 - up to and including version 3.0.x 
* Removed hardcoding for collecting only DeviceInventory or ApplicationInventory. The Azure Function can now handle any log as long as the data is sendt using the correct formatting. 

## Changelog for Inventory Collection Proactive remediation

### 3.5.0 - Requires Azure Function version 1.2 
* Updated output logic in correspondace with new Azure Function returning a payload object instead of string. 
* Fixed a bug in the randomizer function, use the spread the load on the Azure Function over X minutes for larger environment.
* Adding 2 variables
* * Turn on / off randomizer function 
* * Set the timespan to randomize over 
* * Disabled inventory collection during provisioning day.

### 3.0.1 - Requires Azure Function version 1.1 
* Added support for CloudPC (Different method to find AzureAD DeviceID for verification) 

### 3.0.0 -  Requires Azure Function version 1.1 
* Added options to more easily add different custom logs and inventory by creating 1 main payload to be sent to Azure Function
* Added sample PR to show how to ingest your own custom data. 
* Generic bugfixes 

### 2.x - Requires Azure function 1.0 
* Added MACAddress to the inventory for each NIC. 
* Added section to cater for BIOS release version information, for HP, Dell and Lenovo and general bugfixes
* Removed all location gathering for privacy reasons 
* Moved secrets out of code - now running via Azure Function



Add-MSIGraphPermissions.ps1
#Requires -Modules Microsoft.Graph
# Install the module. (You need admin on the machine.)
# Install-Module Microsoft.Graph

# Set Static Variables
$TenantID=""
$ServicePrincipalAppDisplayName =""

# Define dynamic variables
$ServicePrincipalFilter = "displayName eq '$($ServicePrincipalAppDisplayName)'" 
$GraphAPIAppName = "Microsoft Graph"
$ApiServicePrincipalFilter = "displayName eq '$($GraphAPIAppName)'"

# Scopes needed for the managed identity (Add other scopes if needed)
$Scopes = @(
    "Device.Read.All"
)

# Connect to MG Graph - scopes must be consented the first time you run this. 
# Connect with Global Administrator
Select-MgProfile -Name "beta"
Connect-MgGraph -Scopes "Application.Read.All","AppRoleAssignment.ReadWrite.All"  -TenantId $TenantID -UseDeviceAuthentication

# Get the service principal for your managed identity.
$ServicePrincipal = Get-MgServicePrincipal -Filter $ServicePrincipalFilter

# Get the service principal for Microsoft Graph. 
# Result should be AppId 00000003-0000-0000-c000-000000000000
$ApiServicePrincipal = Get-MgServicePrincipal -Filter "$ApiServicePrincipalFilter"

# Apply permissions
Foreach ($Scope in $Scopes) {
    Write-Host "`nGetting App Role '$Scope'"
    $AppRole = $ApiServicePrincipal.AppRoles | Where-Object {$_.Value -eq $Scope -and $_.AllowedMemberTypes -contains "Application"}
    if ($null -eq $AppRole) { Write-Error "Could not find the specified App Role on the Api Service Principal"; continue; }
    if ($AppRole -is [array]) { Write-Error "Multiple App Roles found that match the request"; continue; }
    Write-Host "Found App Role, Id '$($AppRole.Id)'"

    $ExistingRoleAssignment = Get-MgServicePrincipalAppRoleAssignment -ServicePrincipalId $ServicePrincipal.Id | Where-Object { $_.AppRoleId -eq $AppRole.Id }
    if ($null -eq $existingRoleAssignment) {
        New-MgServicePrincipalAppRoleAssignment -ServicePrincipalId $ServicePrincipal.Id -PrincipalId $ServicePrincipal.Id -ResourceId $ApiServicePrincipal.Id -AppRoleId $AppRole.Id
    } else {
        Write-Host "App Role has already been assigned, skipping"
    }
}

Azure Functions/profile.ps1
# Azure Functions profile.ps1
#
# This profile.ps1 will get executed every "cold start" of your Function App.
# "cold start" occurs when:
#
# * A Function App starts up for the very first time
# * A Function App starts up after being de-allocated due to inactivity
#
# You can define helper functions, run commands, or specify environment variables
# NOTE: any variables defined that are not environment variables will get reset after the first execution

# Authenticate with Azure PowerShell using MSI.
# Remove this if you are not planning on using MSI or Azure PowerShell.
<#if ($env:MSI_SECRET) {
    Disable-AzContextAutosave -Scope Process | Out-Null
    Connect-AzAccount -Identity
}#>

# Uncomment the next line to enable legacy AzureRm alias in Azure PowerShell.
# Enable-AzureRmAlias

# You can also define functions or aliases that can be referenced in any of your PowerShell functions.


Azure Functions/LogCollectorAPI/run.ps1
# Intune Enhanced Inventory 
# Version 1.2 
# Created and maintained by @JankeSkanke 
# Requires minimum version  3.5.0 of the Enhanced Inventory Proactive Remediations Script
# Updated 14.Oct.2022

using namespace System.Net
# Input bindings are passed in via param block.
param($Request)
[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
#region functions
function Get-AuthToken {
    <#
    .SYNOPSIS
        Retrieve an access token for the Managed System Identity.
    
    .DESCRIPTION
        Retrieve an access token for the Managed System Identity.
    
    .NOTES
        Author:      Nickolaj Andersen
        Contact:     @NickolajA
        Created:     2021-06-07
        Updated:     2021-06-07
    
        Version history:
        1.0.0 - (2021-06-07) Function created
    #>
    Process {
        # Get Managed Service Identity details from the Azure Functions application settings
        $MSIEndpoint = $env:MSI_ENDPOINT
        $MSISecret = $env:MSI_SECRET

        # Define the required URI and token request params
        $APIVersion = "2017-09-01"
        $ResourceURI = "https://graph.microsoft.com"
        $AuthURI = $MSIEndpoint + "?resource=$($ResourceURI)&api-version=$($APIVersion)"

        # Call resource URI to retrieve access token as Managed Service Identity
        $Response = Invoke-RestMethod -Uri $AuthURI -Method "Get" -Headers @{ "Secret" = "$($MSISecret)" }

        # Construct authentication header to be returned from function
        $AuthenticationHeader = @{
            "Authorization" = "Bearer $($Response.access_token)"
            "ExpiresOn" = $Response.expires_on
        }
        # Handle return value
        return $AuthenticationHeader
    }
}#end function 
function Send-LogAnalyticsData() {
    <#
   .SYNOPSIS
       Send log data to Azure Monitor by using the HTTP Data Collector API
   
   .DESCRIPTION
       Send log data to Azure Monitor by using the HTTP Data Collector API
   
   .NOTES
       Author:      Jan Ketil Skanke
       Contact:     @JankeSkanke
       Created:     2022-01-14
       Updated:     2022-01-14
   
       Version history:
       1.0.0 - (2022-01-14) Function created
   #>
   param(
       [string]$sharedKey,
       [array]$body, 
       [string]$logType,
       [string]$CustomerId
   )
   #Defining method and datatypes
   $method = "POST"
   $contentType = "application/json"
   $resource = "/api/logs"
   $date = [DateTime]::UtcNow.ToString("r")
   $contentLength = $body.Length
   #Construct authorization signature
   $xHeaders = "x-ms-date:" + $date
   $stringToHash = $method + "`n" + $contentLength + "`n" + $contentType + "`n" + $xHeaders + "`n" + $resource
   $bytesToHash = [Text.Encoding]::UTF8.GetBytes($stringToHash)
   $keyBytes = [Convert]::FromBase64String($sharedKey)
   $sha256 = New-Object System.Security.Cryptography.HMACSHA256
   $sha256.Key = $keyBytes
   $calculatedHash = $sha256.ComputeHash($bytesToHash)
   $encodedHash = [Convert]::ToBase64String($calculatedHash)
   $signature = 'SharedKey {0}:{1}' -f $CustomerId, $encodedHash
   
   #Construct uri 
   $uri = "https://" + $CustomerId + ".ods.opinsights.azure.com" + $resource + "?api-version=2016-04-01"
   
   #validate that payload data does not exceed limits
   if ($body.Length -gt (31.9 *1024*1024)){
       throw("Upload payload is too big and exceed the 32Mb limit for a single upload. Please reduce the payload size. Current payload size is: " + ($body.Length/1024/1024).ToString("#.#") + "Mb")
   }
   $payloadsize = ("Upload payload size is " + ($body.Length/1024).ToString("#.#") + "Kb ")
   
   #Create authorization Header
   $headers = @{
       "Authorization"        = $signature;
       "Log-Type"             = $logType;
       "x-ms-date"            = $date;
       "time-generated-field" = $TimeStampField;
   }
   #Sending data to log analytics
   $response = Invoke-WebRequest -Uri $uri -Method $method -ContentType $contentType -Headers $headers -Body $body -UseBasicParsing
   $statusmessage = "$($response.StatusCode):$($payloadsize)"
   return $statusmessage 
}#end function
#endregion functions

Write-Information "LogCollectorAPI function received a request."
#region initialize

# Setting inital Status Code: 
$StatusCode = [HttpStatusCode]::OK

# Define variables from environment
$LogControll = $env:LogControl
# Get secrets from Keyvault
$CustomerId = $env:WorkspaceID
$SharedKey  = $env:SharedKey

# Get TenantID from my logged on MSI account for verification 
$TenantID = $env:TenantID

# Extracting and processing inbound parameters to variables for matching
$MainPayLoad = $Request.Body.LogPayloads
$InboundDeviceID= $Request.Body.AzureADDeviceID
$InboundTenantID = $Request.Body.AzureADTenantID

$LogsReceived = New-Object -TypeName System.Collections.ArrayList
foreach ($Key in $MainPayLoad.Keys) {
    $LogsReceived.Add($($Key)) | Out-Null
}
Write-Information "Logs Received $($LogsReceived)"

#Required empty variable for posting to Log Analytics
$TimeStampField = ""

#endregion initialize

#region script
# Write to the Azure Functions log stream.
Write-Information "Inbound DeviceID $($InboundDeviceID)"
Write-Information "Inbound TenantID $($InboundTenantID)"
Write-Information "Environment TenantID $TenantID"

# Declare response object as Arraylist
$ResponseArray = New-Object -TypeName System.Collections.ArrayList

# Verify request comes from correct tenant
if($TenantID -eq $InboundTenantID){
    Write-Information "Request is comming from correct tenant"
    # Retrieve authentication token
    $Script:AuthToken = Get-AuthToken

    # Query graph for device verification 
    $DeviceURI = "https://graph.microsoft.com/v1.0/devices?`$filter=deviceId eq '$($InboundDeviceID)'"
    $DeviceIDResponse = (Invoke-RestMethod -Method "Get" -Uri $DeviceURI -ContentType "application/json" -Headers $Script:AuthToken -ErrorAction Stop).value

    # Assign to variables for matching 
    $DeviceID = $DeviceIDResponse.deviceId  
    $DeviceEnabled = $DeviceIDResponse.accountEnabled    
    Write-Information "DeviceID $DeviceID"   
    Write-Information "DeviceEnabled: $DeviceEnabled"
    # Verify request comes from a valid device
    if($DeviceID -eq $InboundDeviceID){
        Write-Information "Request is coming from a valid device in Azure AD"
        if($DeviceEnabled -eq "True"){
            Write-Information "Requesting device is not disabled in Azure AD"                       
            foreach ($LogName in $LogsReceived){
                Write-Information "Processing $($LogName)"
                # Check if Log type control is enabled
                if ($LogControll -eq "true"){
                # Verify log name applicability
                Write-Information "Log name control is enabled, verifying log name against allowed values"
                [Array]$AllowedLogNames = $env:AllowedLogNames
                Write-Information "Allowed log names: $($AllowedLogNames)"
                $LogCheck = $AllowedLogNames -match $LogName
                    if(-not ([string]::IsNullOrEmpty($LogCheck))){
                        Write-Host "Log $LogName Allowed"
                        [bool]$LogState = $true
                    }
                    else {
                        Write-Warning "Logname $LogName not allowed"
                        [bool]$LogState = $false
                    }       
                }
                else{
                    Write-Information "Log control is not enabled, continue"
                    [bool]$LogState = $true
                }
                if ($LogState){
                    $Json = $MainPayLoad.$LogName | ConvertTo-Json
                    $LogSize = $json.Length
                    # Verify if log has data before sending to Log Analytics
                    if ($LogSize -gt 0){
                        Write-Information "Log $($logname) has content. Size is $($json.Length)"
                        $LogBody = ([System.Text.Encoding]::UTF8.GetBytes($Json))
                        # Sending logdata to Log Analytics
                        $ResponseLogInventory = Send-LogAnalyticsData -customerId $CustomerId -sharedKey $SharedKey -body $LogBody -logType $LogName
                        Write-Information "$($LogName) Logs sent to LA $($ResponseLogInventory)"
                        $PSObject = [PSCustomObject]@{
                            LogName = $LogName
                            Response = $ResponseLogInventory
                        }
                        $ResponseArray.Add($PSObject) | Out-Null
                        $StatusCode = [HttpStatusCode]::OK
                    }
                    else {
                        # Log is empty - return status 200 but with info about empty log
                        Write-Information "Log $($logname) has no content. Size is $($json.Length)"
                        $PSObject = [PSCustomObject]@{
                            LogName = $LogName
                            Response = "200:Log does not contain data"
                        }
                        $ResponseArray.Add($PSObject) | Out-Null
                    }
                }
                else {
                    Write-Warning "Log $($LogName) is not allowed"
                    $StatusCode = [HttpStatusCode]::OK
                    $PSObject = [PSCustomObject]@{
                        LogName = $LogName
                        Response = "Logtype is not allowed"
                    }
                    $ResponseArray.Add($PSObject) | Out-Null                   
                }
            }
        }
        else{
            Write-Warning "Device is not enabled - Forbidden"
            $StatusCode = [HttpStatusCode]::Forbidden
        }
    }
    else{
        Write-Warning  "Device not in my Tenant - Forbidden"
        $StatusCode = [HttpStatusCode]::Forbidden
    }
}
else{
    Write-Warning "Tenant not allowed - Forbidden"
    $StatusCode = [HttpStatusCode]::Forbidden
}
#endregion script
$body = $ResponseArray | ConvertTo-Json 
# Associate values to output bindings by calling 'Push-OutputBinding'.
Push-OutputBinding -Name Response -Value ([HttpResponseContext]@{
    StatusCode = $StatusCode
    Body = $body
})


Proactive Remediation/Samples/SAMPLE-CustomLogAzureFunction.ps1
<#
.SYNOPSIS
Sample to explain how to call the collector API with your own logdata. The azure function now supports dynamically "any" log data as long as the API call is formatted correctly. 

.DESCRIPTION
Sample to explain how to call the collector API with your own logdata. The azure function now supports dynamically "any" log data as long as the API call is formatted correctly. 

.EXAMPLE
Invoke-CustomLogAzureFunction.ps1 

.NOTES
FileName:    Invoke-CustomLogAzureFunction.ps1 
Author:      Jan Ketil Skanke
Contact:     @JankeSkanke
Created:     2022-22-02
Updated:     2022-22-02

Version history:
1.0.0 - (2022-22-02) Azure Function updated - Requires version 1.1 of Azure Function LogCollectorAPI - 
#>

#region initialize
# Define your azure function URL: 
# Example 'https://<appname>.azurewebsites.net/api/<functioname>'

$AzureFunctionURL = ""

# Enable TLS 1.2 support 
[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
#Control if you want to collect App or Device Inventory or both (True = Collect)

#Set your Log Analytics Log Name
$CustomLog1Name = "MyLog1"
$CustomLog2Name = "MyLog2"
$Date=(Get-Date)
#endregion initialize

#region functions (All functions in here are requried for sample to work)
# Function to get Azure AD DeviceID
function Get-AzureADDeviceID {
    <#
    .SYNOPSIS
        Get the Azure AD device ID from the local device.
    
    .DESCRIPTION
        Get the Azure AD device ID from the local device.
    
    .NOTES
        Author:      Nickolaj Andersen
        Contact:     @NickolajA
        Created:     2021-05-26
        Updated:     2021-05-26
    
        Version history:
        1.0.0 - (2021-05-26) Function created
    #>
	Process {
		# Define Cloud Domain Join information registry path
		$AzureADJoinInfoRegistryKeyPath = "HKLM:\SYSTEM\CurrentControlSet\Control\CloudDomainJoin\JoinInfo"
		
		# Retrieve the child key name that is the thumbprint of the machine certificate containing the device identifier guid
		$AzureADJoinInfoThumbprint = Get-ChildItem -Path $AzureADJoinInfoRegistryKeyPath | Select-Object -ExpandProperty "PSChildName"
		if ($AzureADJoinInfoThumbprint -ne $null) {
			# Retrieve the machine certificate based on thumbprint from registry key
			$AzureADJoinCertificate = Get-ChildItem -Path "Cert:\LocalMachine\My" -Recurse | Where-Object { $PSItem.Thumbprint -eq $AzureADJoinInfoThumbprint }
			if ($AzureADJoinCertificate -ne $null) {
				# Determine the device identifier from the subject name
				$AzureADDeviceID = ($AzureADJoinCertificate | Select-Object -ExpandProperty "Subject") -replace "CN=", ""
				# Handle return value
				return $AzureADDeviceID
			}
		}
	}
} #endfunction 
function Get-AzureADJoinDate {
    <#
    .SYNOPSIS
        Get the Azure AD device ID from the local device.
    
    .DESCRIPTION
        Get the Azure AD device ID from the local device.
    
    .NOTES
        Author:      Nickolaj Andersen
        Contact:     @NickolajA
        Created:     2021-05-26
        Updated:     2021-05-26
    
        Version history:
        1.0.0 - (2021-05-26) Function created
    #>
	Process {
		# Define Cloud Domain Join information registry path
		$AzureADJoinInfoRegistryKeyPath = "HKLM:\SYSTEM\CurrentControlSet\Control\CloudDomainJoin\JoinInfo"
		
		# Retrieve the child key name that is the thumbprint of the machine certificate containing the device identifier guid
		$AzureADJoinInfoThumbprint = Get-ChildItem -Path $AzureADJoinInfoRegistryKeyPath | Select-Object -ExpandProperty "PSChildName"
		if ($AzureADJoinInfoThumbprint -ne $null) {
			# Retrieve the machine certificate based on thumbprint from registry key
			$AzureADJoinCertificate = Get-ChildItem -Path "Cert:\LocalMachine\My" -Recurse | Where-Object { $PSItem.Thumbprint -eq $AzureADJoinInfoThumbprint }
			if ($AzureADJoinCertificate -ne $null) {
				# Determine the device identifier from the subject name
				$AzureADJoinDate = ($AzureADJoinCertificate | Select-Object -ExpandProperty "NotBefore") 
				# Handle return value
				return $AzureADJoinDate
			}
		}
	}
} #endfunction 
#Function to get AzureAD TenantID
function Get-AzureADTenantID {
	# Cloud Join information registry path
	$AzureADTenantInfoRegistryKeyPath = "HKLM:\SYSTEM\CurrentControlSet\Control\CloudDomainJoin\TenantInfo"
	# Retrieve the child key name that is the tenant id for AzureAD
	$AzureADTenantID = Get-ChildItem -Path $AzureADTenantInfoRegistryKeyPath | Select-Object -ExpandProperty "PSChildName"
	return $AzureADTenantID
}                          
#endregion functions

#region script
#Get Common data for validation in Azure Function: 
$AzureADDeviceID = Get-AzureADDeviceID
$AzureADTenantID = Get-AzureADTenantID

#region CustomLogData1

#Gather some data 

$MyLog1Data1 = "TestData"
$MyLog1Data2 = "TestData"
$MyLog1Data3 = "TestData"
$MyLog1Data4 = "TestData"

# Create Object to Upload to Log Analytics
$Inventory = New-Object System.Object
$Inventory | Add-Member -MemberType NoteProperty -Name "AzureADDeviceID" -Value "$AzureADDeviceID" -Force
$Inventory | Add-Member -MemberType NoteProperty -Name "MyLog1Data1" -Value "$MyLog1Data1" -Force
$Inventory | Add-Member -MemberType NoteProperty -Name "MyLog1Data2" -Value "$MyLog1Data2" -Force
$Inventory | Add-Member -MemberType NoteProperty -Name "MyLog1Data3" -Value "$MyLog1Data3" -Force
$Inventory | Add-Member -MemberType NoteProperty -Name "MyLog1Data4" -Value "$MyLog1Data4" -Force
$MyLog1DataInventory = $Inventory
	
#endregion CustomLogData1

#region CustomLogData2
	
# Data Arrays are also supported 

$MyArray = "Query an array of data"


$MyArray = @()
foreach ($Item in $MyArray) {
	$tempData = New-Object -TypeName PSObject
	$tempData | Add-Member -MemberType NoteProperty -Name "AzureADDeviceID" -Value "$AzureADDeviceID" -Force
	$tempData | Add-Member -MemberType NoteProperty -Name "MyLog2Data1" -Value $Item.MyLog2Data1 -Force
	$tempData | Add-Member -MemberType NoteProperty -Name "MyLog2Data2" -Value $Item.MyLog2Data2 -Force
	$tempData | Add-Member -MemberType NoteProperty -Name "MyLog2Data3" -Value $Item.MyLog2Data3 -Force 
	$tempData | Add-Member -MemberType NoteProperty -Name "MyLog2Data4" -Value $Item.MyLog2Data4 -Force
	$DataArray += $tempData
}

$MyLog2DataInventory = $DataArray

#endregion CustomLogData2

#Randomize over 50 minutes to spread load on Azure Function - disabled on date of enrollment (Disabled in sample - Enable only in larger environments)
$JoinDate = Get-AzureADJoinDate
$DelayDate = $JoinDate.AddDays(1)
$CompareDate = ($DelayDate - $JoinDate)
if ($CompareDate.Days -ge 1){
	Write-Output "Randomzing execution time"
	#$ExecuteInSeconds = (Get-Random -Maximum 3000 -Minimum 1)
	#Start-Sleep -Seconds $ExecuteInSeconds
}
#Start sending logs
$date = Get-Date -Format "dd-MM HH:mm"
$OutputMessage = "InventoryDate:$date "

$headers = New-Object "System.Collections.Generic.Dictionary[[String],[String]]"
$headers.Add("Content-Type", "application/json")

# Add arrays of logs into payload array 
$LogPayLoad = New-Object -TypeName PSObject 
$LogPayLoad | Add-Member -NotePropertyMembers @{$CustomLog1Name= $MyLog1DataInventory}
$LogPayLoad | Add-Member -NotePropertyMembers @{$CustomLog2Name = $MyLog2DataInventory}

# Construct main payload to send to LogCollectorAPI // IMPORTANT // KEEP AND DO NOT CHANGE THIS
$MainPayLoad = [PSCustomObject]@{
	AzureADTenantID = $AzureADTenantID
	AzureADDeviceID = $AzureADDeviceID
	LogPayloads = $LogPayLoad
}
$MainPayLoadJson = $MainPayLoad| ConvertTo-Json -Depth 9	

# Sending data to API
try {
	$ResponseInventory = Invoke-RestMethod $AzureFunctionURL -Method 'POST' -Headers $headers -Body $MainPayLoadJson
	$OutputMessage = $OutPutMessage + "Inventory:OK " + $ResponseInventory
} 
catch {
	$ResponseInventory = "Error Code: $($_.Exception.Response.StatusCode.value__)"
	$ResponseMessage = $_.Exception.Message
	$OutputMessage = $OutPutMessage + "Inventory:Fail " + $ResponseInventory + $ResponseMessage
}

# Check status and report to Proactive Remediations
if ($ResponseInventory-match "200"){
	Write-Output $OutputMessage
	Exit 0
} else {
	Write-Output "Error: $($ResponseInventory), Message: $($ResponseMessage)"
	Exit 1
	}
#endregion script
